### PR TITLE
fixes #22388 - jwt token auth

### PR DESCRIPTION
**Pull request by timogoebel**
This is part 1 of the graphql series and extracts JWT auth from #5336. It also adds JWT as an SSO method.

Currently there is no API route to return a jwt token for a user. I'd like to add that in a separate PR.

Credits go to @dariuszb-iRonin and @lucasm-iRonin.

---
Other parts:
Part 1 - JWT Auth: #5596
Part 2 - Graphql Scaffolding: #5680
Part 3 - Relay Global ID #5681
Part 4 - Connections with totalCount #5733
Part 5 - Basic mutations for model #5736

### CODE REVIEW

**Comment by timogoebel** [comment]
[test foreman]
[test upgrade]

**Comment by timogoebel** [comment]
[test upgrade]

File: app/models/jwt_secret.rb
Comment (leewaa) [inline app/models/jwt_secret.rb]: A scope would be nicer instead of the class method:
```ruby
scope :for_user ->(user) { find_by(user: user) }
```
Also I would use the Argument name `user` instead of `user_or_user_id` since its a standard rails feature being able to pump in either the object or id for querying by a relation so there is no added value with such naming.

However, you should even consider removing this all together and just sticking with `JwtSecret.find_by(user: user)`. Cleaner imho. I see you do stub this method, however you can just as easily stub the `find_by` call.

File: app/models/jwt_token.rb
Comment (leewaa) [inline app/models/jwt_token.rb]: the trailing `.to_s` is redundant or not?

File: test/models/jwt_token_test.rb
Comment (leewaa) [inline test/models/jwt_token_test.rb]: So as per my comment above, you can do:
```ruby
JwtSecret.stubs(:find_by).returns(jwt_secret)
```

**Comment by timogoebel** [comment]
@leewaa: Thanks. I addressed your comments but decided to keep the scope for now. :-)

**Comment by martinpovolny** [comment]
I am trying to understand why you need the `JwtSecret` model here and so far I am failing at that.

As I understand it you should decode the token and the information in the token should be trusted. That's the point of JWT. You should not need to lookup the token in the database.

Sure, if the token contains a user id, you lookup that user. But not the token.

This PR seems to be using JWT the same way you would use a session cookie.

I may be missing something or course.

What do you think @timogoebel ?

**Comment by timogoebel** [comment]
@martinpovolny:

> As I understand it you should decode the token and the information in the token should be trusted. That's the point of JWT. You should not need to lookup the token in the database.

That's completely true. In order to verify that the token is valid and can be trusted the token is signed using a secret. That secret has to be stored somewhere to verify the token. This PR stores a token that is unique per user to increase security in a model called `JwtSecret`. In the future, we can leverage this to invalidate all tokens for a specific user in case a token get's lost or something.
I hope, I could shed some light.

**Comment by martinpovolny** [comment]
@timogoebel : Thx for the explanation. 

Ok, that seems a valid reason that makes sense as you explain it. Although I suspect that it defeats part of the purpose for JWT.

It might not be relevant for Foreman or not right now. But as you wrote in the issue https://projects.theforeman.org/issues/22388 :

> JWTs have the advantage, that they are stateless, scalable and decoupled. They also promise better performance and are mobile ready.

I understand it that with the use of JWT in a more advanced architecture (proxies, load balancers, microservices) and as written in https://jwt.io/introduction/ :

> The server's protected routes will check for a valid JWT in the Authorization header, and if it's present, the user will be allowed to access protected resources.

Meaning that you don't need to query the database to check the token.

A load balancer for example can check the authorization w/o being able to talk to a database or a memcache or something. Just by verifying the token.

A micro service can have all the authorization information in the token so it does not need to talk to a database to trust a user and perform a task.

I know that this is not very relevant with current Foreman architecture but in when there's a SPA talking to a number of APIs (and I guess that is the future) the situation becomes different.

In your implementation one needs to talk to the DB to verify the token. In your case the state is the encryption key and there's no decoupling.

I think your approach works perfectly well for current Foreman. But I also think that it's not the best way to do JWT.

But I have to confess that my practical experience with this topic is very shallow.

**Comment by timogoebel** [comment]
[test foreman]

**Comment by ohadlevy** [comment]
@martinpovolny thats indeed interesting, how does one de-validate a token in cases its used by LB, proxies etc?

**Comment by martinpovolny** [comment]
@ohadlevy : there's no "nice" mechanism for that in JWT (at least to my understanding). But there's a nice stackoverflow question that, I think, has everything in it:

https://stackoverflow.com/questions/21978658/invalidating-json-web-tokens

The ultimate solution is in the second answer:

> .... but a very simple and easy way to invalidate all the existing JWTs is simply to change the secret.

File: test/models/user_test.rb
Comment (dariuszb-iRonin) [inline test/models/user_test.rb]: This test is duplicated

File: test/factories/user_related.rb
Comment (dariuszb-iRonin) [inline test/factories/user_related.rb]: We do not use that anywhere. Do we want to have it for the future?

**Review by dariuszb-iRonin — COMMENTED** [review]
Looks very good to me. I just added two small comments

**Comment by timogoebel** [comment]
@dariuszb-iRonin: Thanks, I have addressed your comments and rebased the PR.

File: app/models/jwt_token.rb
Comment (ik5) [inline app/models/jwt_token.rb]: what will happen if there is an exception here?

File: app/models/jwt_token.rb
Comment (ik5) [inline app/models/jwt_token.rb]: why md5 and not sha2 for example?
If my memory serves me right, the default is sha2-256

File: app/models/jwt_token.rb
Comment (ik5) [inline app/models/jwt_token.rb]: why do you return here `nil`, and on the usage you are converting it into empty hash?

**Review by ik5 — COMMENTED** [review]
few comments and questions regarding the code, nothing really blocking

File: app/models/jwt_token.rb
Comment (timogoebel) [reply app/models/jwt_token.rb]: ... to make this self-contained.

File: app/models/jwt_token.rb
Comment (timogoebel) [reply app/models/jwt_token.rb]: I don't think we need to worry about this. `find_by` returns `nil` if the entry does not exist.

File: app/models/jwt_token.rb
Comment (timogoebel) [inline app/models/jwt_token.rb]: Can this be done with the library?

File: app/services/sso/jwt.rb
Comment (timogoebel) [inline app/services/sso/jwt.rb:10]: Let's double check this actually checks the JWT signature matches the signature of the payload.

File: app/models/jwt_token.rb
Comment (timogoebel) [reply app/models/jwt_token.rb]: Actually, this is because we don't want to check the signature here. The library supports this use case, so we should use it here.

**Comment by xprazak2** [comment]
Visually looking ok, needs a couple of minor changes based on previous comments. It would be good to test together with API to verify it works as expected.

File: Gemfile
Comment (tbrisker) [inline Gemfile]: this was extracted?

File: config/initializers/cors.rb
Comment (tbrisker) [inline config/initializers/cors.rb]: also this file should be removed?

File: test/unit/sso/jwt_test.rb
Comment (tbrisker) [inline test/unit/sso/jwt_test.rb:19]: shouldn't the expectation come first?

File: app/services/sso/jwt.rb
Comment (tbrisker) [reply app/services/sso/jwt.rb:10]: can we add a test for this - making sure one user's secret can't be used to encode another user's id?

**Review by tbrisker — CHANGES_REQUESTED** [review]
some comments inline, also don't forget to feed the dog ;)

File: app/services/sso/jwt.rb
Comment (timogoebel) [reply app/services/sso/jwt.rb:10]: @kamils-iRonin: Can you help me with writing the test? I don't find the time to do it myself. Thank you.

**Comment by timogoebel** [comment]
I believe I addressed all comments. @kamils-iRonin is working on adding the requested test.

File: app/models/jwt_secret.rb
Comment (kamils-iRonin) [reply app/models/jwt_secret.rb]: @timogoebel scope `for_user` returns all records if there is no record matching the user
```
[1] pry(main)> JwtSecret.for_user(User.find(1))
=> [#<JwtSecret:0x00007fded3df2190
  id: 1,
  token: "rmZbjJV3RIOQ5shPb21COQ==",
  user_id: 2,
  created_at: Tue, 12 Feb 2019 14:07:25 UTC +00:00,
  updated_at: Tue, 12 Feb 2019 14:07:25 UTC +00:00>,
 #<JwtSecret:0x00007fded3df2000
  id: 4,
  token: "8LsZmPrAwtOcu/Zi3q3idQ==",
  user_id: 5,
  created_at: Tue, 12 Feb 2019 14:09:39 UTC +00:00,
  updated_at: Tue, 12 Feb 2019 14:09:39 UTC +00:00>]
```
I suggest to go with `find_by` instead
```
[2] pry(main)> JwtSecret.find_by(user: User.find(1))
=> nil
```
If we want to keep the scope we should use `where` instead of `find_by` in the scope as scope should return `ActiveRecord::Relation` object.

**Comment by timogoebel** [comment]
I incorporated @kamils-iRonin's changes: https://github.com/kamils-iRonin/foreman/commit/025aa2aa5b39d9dcf23017ba70d69ae2486fccab

File: app/models/concerns/jwt_auth.rb
Comment (tbrisker) [inline app/models/concerns/jwt_auth.rb:8]: Where is `create_jwt_secret!` defined?

**Review by tbrisker — COMMENTED** [review]
one last question, otherwise lgtm!

File: app/models/concerns/jwt_auth.rb
Comment (timogoebel) [reply app/models/concerns/jwt_auth.rb:8]: It's rails magic, see https://guides.rubyonrails.org/association_basics.html#methods-added-by-has-one.

Comment (tbrisker) [reply app/models/concerns/jwt_auth.rb:8]: oooh, magic! :tophat: :rabbit:

**Review by tbrisker — APPROVED** [review]
Thanks  @dariuszb-iRonin, @lucasm-iRonin, @timogoebel, @kamils-iRonin ! any other comments @xprazak2 ?

**Comment by xprazak2** [comment]
I am good.

File: app/models/jwt_secret.rb
Comment (ohadlevy) [inline app/models/jwt_secret.rb:10]: is there a notion of token expiry based on time? I wonder if we can use jwt with the new login page, but not sure how to maintain session expiry

Comment (timogoebel) [reply app/models/jwt_secret.rb:10]: Yeah, jwt tokens support expiry. We can add that later. This PR just allows users to sign in with tokens but tokens aren't issued anywhere right now. That's something we need to add in a follow up PR.

**Comment by timogoebel** [comment]
@xprazak2, @tbrisker: I dare you to hit the merge button. :-)

**Comment by tbrisker** [comment]
Looks like code owners just stopped me from breaking nightlies :) we already have packaging for jwt, but we ship 1.2.1, while this PR requires ~>2.1. AFAICT, we use jwt in the proxy as well, so would be good to verify this upgrade doesn't break it.

**Comment by timogoebel** [comment]
@tbrisker: We need the older version for the proxy to support ancient ruby versions. In core we want and need a more recent version of the library. No need to block this PR any further.

**Comment by tbrisker** [comment]
@timogoebel @ekohl can we ship both versions? right now we don't have 2.1 packaged, so i think nightlies will fail once this is merged.

**Comment by timogoebel** [comment]
@ekohl, @evgeni: How can we proceed 📦-ing wise here?

**Comment by evgeni** [comment]
Without looking too closely, I think we need to split the packaging into a modern/scl(tfm) version and one for the proxy. I think we already do this for the restclient gem, so it shouldn't be too complicated to achieve.

**Comment by timogoebel** [comment]
@evgeni: Thanks for your swift rely. Is that something you can help me out with?

**Comment by evgeni** [comment]
Not this weekend, but generally yes

**Comment by evgeni** [comment]
Just to be sure, core currently doesn't use the gem at all, right? So we don't need to ensure that rest of core works fine with the newer version when we adjust packaging

**Comment by timogoebel** [comment]
> Not this weekend, but generally yes

Obviously, I am just bored at a conference. ;-)

> Just to be sure, core currently doesn't use the gem at all, right?

I was wondering the same and couldn't figure out why the package was scled in the first place.

I tried to do this myself, please see https://github.com/theforeman/foreman-packaging/pull/3466

**Comment by timogoebel** [comment]
@evgeni: :  📦🕐🚢?

**Review by evgeni — APPROVED** [review]
:package: :+1: :bangbang: :nail_care:

**Comment by tbrisker** [comment]
Thanks @timogoebel, @evgeni and everyone else involved in this!

File: app/models/jwt_secret.rb
Comment (rabajaj0509) [reply app/models/jwt_secret.rb:10]: I am working on SSO integration with foreman via keycloak, I will take care of the token expiry and related things :)

Comment (timogoebel) [reply app/models/jwt_secret.rb:10]: Excellent, as a quick hint: This should do the trick:

https://github.com/timogoebel/foreman_register/blob/d7acbed10f9ec9c85891ec8823bf01b70f607d8a/app/services/foreman_register/registration_token.rb#L13-L26

**Pull request merged**
